x-pack/plugins/data_enhanced/public/search/search_response_cache.ts

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { Observable, Subscription } from 'rxjs';
+import { IKibanaSearchResponse, isErrorResponse } from '../../../../../src/plugins/data/public';
+import { SearchAbortController } from './search_abort_controller';
+
+interface ResponseCacheItem {
+  response$: Observable<IKibanaSearchResponse<any>>;
+  searchAbortController: SearchAbortController;
+}
+
+interface ResponseCacheItemInternal {
+  response$: Observable<IKibanaSearchResponse<any>>;
+  searchAbortController: SearchAbortController;
+  size: number;
+  subs: Subscription;
+}
+
+export const CACHE_MAX_SIZE_MB = 10;
+
+export class SearchResponseCache {
+  private responseCache: Map<string, ResponseCacheItemInternal>;
+  private cacheSize = 0;
+
+  constructor(private maxItems: number, private maxCacheSizeMB: number) {
+    this.responseCache = new Map();
+  }
+
+  private byteToMb(size: number) {
+    return size / (1024 * 1024);
+  }
+
+  private deleteItem(key: string, clearSubs = true) {
+    const item = this.responseCache.get(key);
+    if (item) {
+      if (clearSubs) {
+        item.subs.unsubscribe();
+      }
+      this.cacheSize -= item.size;
+      this.responseCache.delete(key);
+    }
+  }
+
+  private setItem(key: string, item: ResponseCacheItemInternal) {
+    // The deletion of the key will move it to the end of the Map's entries.
+    this.deleteItem(key, false);
+    this.cacheSize += item.size;
+    this.responseCache.set(key, item);
+  }
+
+  public clear() {
+    this.cacheSize = 0;
+    this.responseCache.forEach((item) => {
+      item.subs.unsubscribe();
+    });
+    this.responseCache.clear();
+  }
+
+  private shrink() {
+    while (
+      this.responseCache.size > this.maxItems ||
+      this.byteToMb(this.cacheSize) > this.maxCacheSizeMB
+    ) {
+      const [key] = this.responseCache.entries().next().value as [string, ResponseCacheItem];
```

Comment:
Nit: You can remove the cast
```suggestion
      const [key] = [...this.responseCache.keys()];
```